File: K-diffPairsInArray.java
class Solution {
    public int findPairs(int[] nums, int k) {
       Arrays.sort(nums); // trick sort first to not get negatives 
      
       int n = nums.length; 
       int slow = 0;
       int fast = 0;
       int count = 0;
       while(slow < n && fast < n){
          
           //case1: First check slow and fast same or not 
           if( slow == fast ||  nums[fast] - nums[slow] < k )  fast++; //bring slow closer to fast pointer
           
           //case2:
           else if(nums[fast]- nums[slow] > k) slow++;  
           
           //case3
           else{
               count++;
               slow++;
               fast++;   // no need to take another ds like list and count them because we dont want pairs we only want count
           
               while(fast < n &&  nums[fast] == nums[fast -1]) fast++; //also dont forget to check if prev and current val same increment fast
           }    
       }
        return count;
    } 
}

/*
TimeComplexity: O(nlogn)(sort) + O(n)(traverse each element) = O(nlogn)
SpaceComplexity: O(1) as used same nums array for computation
*/

File: KDiffPairsInArray.py
#TC: O(n log(n)), n: length of nums
#SC: O(1)
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        slow = 0
        fast = 1
        count = 0
        nums.sort()
        while slow < len(nums) and fast < len(nums):
            diff = nums[fast] - nums[slow]
            if diff == k:
                count+=1
                slow += 1
                fast += 1
                while fast < len(nums) and nums[fast]==nums[fast-1]:
                    fast += 1
            elif diff > k:
                slow+=1
            else:
                fast+=1
            if slow == fast:
                fast += 1

        return count


